### PR TITLE
TSC Graph Context: handle property access

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, the `tsc-mixed` context didn't boost the type of the *expression* (aka. qualifier) in property accesses. For example, the property access
```ts
this.helper().CURSOR
```
has the *expression* `this.helper()`, which might have the type `interface Helper { ... }` with relevant members for autocomplete. This PR fixes the problem by effectively doing "Go to type definition" on the expression and including that definition in the context.

Fixes CODY-1331

## Test plan

Green CI. See newly added tests.

### CODE REVIEW

File: agent/src/graph-context.test.ts [hunk 1]
```diff
@@ -541,6 +541,189 @@ describe.skipIf(isWindows())('Graph Context', () => {
...
+            import { getter } from './members-indirection'
+
+            function run(): void {
+                getter.indirect()./* CURSOR */
+            }
+            `
+            )
+
+            const text = await autocompletes()
+            // expect(text).includes('validDogSled')
```
Comment (philipp-spiess) [inline agent/src/graph-context.test.ts:616]:
```suggestion
```

File: agent/src/graph-context.test.ts [hunk 2]
```diff
@@ -541,6 +541,189 @@ describe.skipIf(isWindows())('Graph Context', () => {
...
+
+                        //
+
+                        // interface Selector {
+
+                        //   query(params: { isMammal: boolean; animalName: string; }) => { animals: Animal[]; }
+
+                        // }
+
+                        //
```
Comment (philipp-spiess) [inline agent/src/graph-context.test.ts:647]: Still think there's something wrong with how these are formatted. They seem to be terminated at with two `\n\n` 🤔  The code below doesn't seem to do that. I have commented that before but I don't remember if/what your response was. 😐

**Review by philipp-spiess — APPROVED** [review]
Really cool!

File: agent/src/graph-context.test.ts (hunk 2 above)
Comment (olafurpg) [reply agent/src/graph-context.test.ts:825]: I think there's a bug in our prompt logic that inserts these blank lines, you can see it in the recordings

https://sourcegraph.com/github.com/sourcegraph/cody@6b39919a9c7242f51f7c5f4cf73e14abc1b07f08/-/blob/agent/recordings/graph-context-starcoder-16b_352709139/recording.har.yaml?L180-197

CleanShot 2024-05-03 at 11 19 07@2x


I was planning to comment on this in Slack. I was reading the `PromptString.join()` logic and I think it's currently difficult to reason about, and I am concerned about the performance of repeated usage of split/join (prompts can become big strings).

Comment (olafurpg) [reply agent/src/graph-context.test.ts:823]: [CODY-1364 : Blank lines in autocomplete context prompt](https://linear.app/sourcegraph/issue/CODY-1364/blank-lines-in-autocomplete-context-prompt)

Comment (philipp-spiess) [reply agent/src/graph-context.test.ts:823]: Yeah, testing the fix right now. Regarding the performance overhead: Agree and it's something we may want to look into but we should first gather some data to see how big of a problem this really is before we spent time (we'd need to change the datatype for how we store references so we can concatenate faster).

Comment (olafurpg) [reply agent/src/graph-context.test.ts:823]: +1 on sampling data for perf overhead for prompt construction

**Comment by olafurpg** [comment]
Failing ubuntu 18 tests are unrelated to this PR

**Pull request merged**